Fast ICM42xxx performance

andyp1per opened the pull request:
At faster gyro backend rates (e.g. 4khz) the per-tick CPU load becomes quite important. This change reduces the CPU load a couple of ways:

- Increases the SPI clock on the MatekH743 to 16Mhz for the ICM42xxx gyros which are rated at up to 24Mhz (cannot currently get higher than 12.5Mhz because of the clock divisor). This gives about a 30% reduction in per-tick time.
- Synchronizes the ticks with the data. Before there were a lot of redundant ticks because data wasn't quite ready. Now the next tick is calculated from the time at which data is received.

timtuxworth commented:
Would this also apply for the Zealot H743 which has a ICM-42688-P?

andyp1per commented:
> Would this also apply for the Zealot H743 which has a ICM-42688-P

Yes, but manufacturer would need to confirm it is safe

timtuxworth commented:
> > Would this also apply for the Zealot H743 which has a ICM-42688-P
> 
> Yes, but manufacturer would need to confirm it is safe

I've checked with QIo-Tek, they say this fix isn't necessary because their design doesn't have the problem this is fixing.

andyp1per commented:
> I've checked with QIo-Tek, they say this fix isn't necessary because their design doesn't have the problem this is fixing.

Which can only mean they are not using this chip

QioTek commented:
> > I've checked with QIo-Tek, they say this fix isn't necessary because their design doesn't have the problem this is fixing.
> 
> Which can only mean they are not using this chip

It needs to be tested. Will higher sampling make the sampling accuracy of the sensor different? We haven't conducted actual hardware testing yet. We need to compare with the previous sampling speed.

andyp1per commented:
> It needs to be tested. Will higher sampling make the sampling accuracy of the sensor different? We haven't conducted actual hardware testing yet. We need to compare with the previous sampling speed.

For the 42688 the sampling is all done at 32Khz regardless of ODR. Higher ODR allows higher AP filter settings. This issue is addressing some issues that might be seen with higher ODR - mostly because the tick latency becomes more important

andyp1per commented:
Flew four batteries on this today - all worked great

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp:
```diff
@@ -119,7 +119,7 @@ struct PACKED FIFOData {
 
 #define INV3_SAMPLE_SIZE sizeof(FIFOData)
-#define INV3_FIFO_BUFFER_LEN 8
+#define INV3_FIFO_BUFFER_LEN 2
```
i'd be worried about getting behind

andyp1per replied:
Fixed

tridge commented:
tested on Pixhawk6X, works well!
